types/react-redux/react-redux-tests.tsx

```diff
@@ -102,23 +102,79 @@ function MapDispatch() {
 }
 
 function MapDispatchWithThunkActionCreators() {
-    class TestComponent extends React.Component<{
-        foo: string,
-        onClick(): void,
-        thunkAction(): Promise<void>
-    }> {}
-
-    const mapDispatchToProps = () => ({
-        onClick: () => {},
-        thunkAction: () => async () => {}
+    const simpleAction = (payload: boolean) => ({
+        type: 'SIMPLE_ACTION',
+        payload,
     });
-
-    const Test = connect(
-        null,
-        mapDispatchToProps,
+    const thunkAction = (param1: number, param2: string) => (
+        async (dispatch: Dispatch, { foo }: OwnProps) => {
+            return foo;
+        }
+    );
+    interface OwnProps {
+        foo: string;
+    }
+    interface TestComponentProps extends OwnProps {
+        simpleAction: typeof simpleAction;
+        thunkAction(param1: number, param2: string): Promise<string>;
+   }
+    class TestComponent extends React.Component<TestComponentProps> {}
+
+    const mapStateToProps = ({ foo }: { foo: string }) => ({ foo });
+    const mapDispatchToProps = { simpleAction, thunkAction };
+
+    const Test1 = connect(null, mapDispatchToProps)(TestComponent);
+    const Test2 = connect(mapStateToProps, mapDispatchToProps)(TestComponent);
+    const Test3 = connect(
+        null, mapDispatchToProps, null, { storeKey: 'somekey' }
     )(TestComponent);
+    const Test4 = connect(
+        mapStateToProps, mapDispatchToProps, null, { storeKey: 'somekey' }
+    )(TestComponent);
+    const verify = <div>
+        <Test1 foo='bar' />;
+        <Test2 />
+        <Test3 foo='bar' />;
+        <Test4 />
+    </div>;
+}
 
-    const verify = <Test foo='bar' />;
+function MapManualDispatchThatLooksLikeThunk() {
+    interface OwnProps {
+        foo: string;
+    }
+    interface TestComponentProps extends OwnProps {
+        remove: (item: string) => () => object;
+   }
+    class TestComponent extends React.Component<TestComponentProps> {
+        render() {
+            return <div onClick={this.props.remove('someid')} />;
+        }
+    }
+
+    const mapStateToProps = ({ foo }: { foo: string }) => ({ foo });
+    function mapDispatchToProps(dispatch: Dispatch) {
+        return {
+            remove(item: string) {
+                return () => dispatch({ type: 'REMOVE_ITEM', item });
```

Comment:
If you're using `redux-thunk` you should be using `ThunkDispatch` as the type of `dispatch` which would correctly handle thunk action creators.

Probably better to add a test just to make sure `mapDispatchToProps(dispatch: ThunkDispatch)` is assignable as the 2nd argument.

Reply:
Can I use a `ThunkDispatch` type in tests without adding explicit dependency on `redux-thunk`?

If "No", then what are the benefits  of using `ThunkDispatch` instead of simple `Dispatch` in the test? As far as I can tell `react-redux` will NOT use type inference to pick up `ThunkDispatch` type from `mapDispatchToProps` argument (i.e. `react-redux` will still use simple `Dispatch` under the hood  regardless of what type you're using in test).

I'd like to avoid direct explicit dependency on `redux-thunk`, because:
- `react-redux` can easily be used without `redux-thunk`, so it's there's no actual depenendency in place (and I'm not sure if adding `redux-thunk` to `react-redux`'s  `package.json` will not lead to problem for the people not using `redux-thunk`)
- even if we use `redux-thunks` only for the tests, future maintainers may use it in the type definitions (just because the dependency is already there), which may make them less generic (they will stop working for custom `redux-thunk` implemenations, for example)
- it will be a bad example for the future maintainers that may add dependencies on other Redux middlewares by analogy which may affect `redux-thunk`'s definitions verstaility/maintainability (I cannot imagine how though; it more like a gut feeling).

Reply:
Not really...... would probably have to copy the definition

```ts
export interface ThunkDispatch<S, E, A extends Action> {
  <T extends A>(action: T): T;
  <R>(asyncAction: ThunkAction<R, S, E, A>): R;
}
```

Because it doesn't `extends Dispatch` it looks like it wouldn't work, not directly. With the way the `Middleware` / `StoreEnhancer` interface works, they actually can upgrade the type of `dispatch` in a specific `redux` store (which would have to be accessed with `typeof store['dispatch']`), but I'm not sure there's any way to propagate it to `react-redux` as `connect` doesn't know what store it is connecting to.

So there are two alternatives: do something similar to what `styled-components` does and have a way to reexport in your own code a version of `react-redux` where the store/dispatch types are hardcoded to types you're aware of, or change `redux-thunk` to directly augment `Dispatch<A>` instead.

Or recommend that users write the module augmentation themselves:

```ts
import { ThunkDispatch } from 'redux-thunk'

interface State {
  // your redux store state
}

declare module 'redux' {
  interface Dispatch<A extends Action> extends ThunkDispatch<State, never, A> {}
}
```

Reply:
Thank you very much for explanation, but to be honest I did not get the complete picture out of it (seems like I'm not yet ready for such an abstract types; I'll probably need to re-read your comment again and again to understand it).

Even if the problem is real (I'll check if it exists for me in my project), it seems like it was there even before I introduced any of my thunk-changes - hence it's not immediate subject of this PR. I'm going to contemplate on it and maybe I'll come up with the follow up PR to address it.

Right now I'm more concerned by the fact that my previous PR introduced regression into the @types/react-reduc@6.0.10, so I'd like to merge this PR as fast as possible to fix previously introduced regression.

So, if you don't mind, let's delay this problem for a while.